uni/lib/FxPt.icn

```diff
@@ -0,0 +1,627 @@
+         if p1 > p then { # Same external value, more decimal places
+            return _FxPt(integer(v * 10^(p1-p)), p1)
+         }
+         else {
+            v ? {
+               v1 := tab(p1-p)
+               if *v1 = 0 then return _FxPt(0,p1)
+               if move(1) >= 5 then { if v > 0 then v1 +:=1 else v1 -:=1 }
+               return _FxPt(integer(v1), p1)
+            }
+         }
+      }
+      # fail
+   end
+
+   method trunc(p1)  # truncate toward zero to p1 decimal places
+      local v1
+      dbgPV { "Truncating", toString(), "to", p1, "places" }
+       if p1 >= 0 then {
+        if p1 = p then {return self} # Nothing to do
+         if p1 > p then {
+            return _FxPt(integer(v * 10^(p1-p)), p1)
+         }
+         else {
+            v ? {
+               v1 := tab(p1-p)
+               if *v1 = 0 then return _FxPt(0,p1)
+               return _FxPt(integer(v1), p1)
+            }
+         }
+      }
+      # fail
+   end
+
+   # Internal method: compare self with fp (return -1, 0, 1  for <, =, >)
+   method _cmp(fp)
+      local myV, fpV
+      if fp.p = p then { # easy
+         myV := integer(v); fpV := integer(fp.v)
+       } else {
+         if p > fp.p then {
+            myV := integer(v); fpV := integer(fp.v||repl("0",p-fp.p))
+         } else {
+            myV := integer(v || repl("0",fp.p-p)); fpV := integer(fp.v)
+         }
+      }
+      return if myV < fpV then -1 else if myV > fpV then 1 else 0
+   end
+
+   method eq(fp)  # self = fp?
+      return 0 = _cmp(fp)
+   end
+
+   method ne(fp)  # self ~= fp?
+      return 0 ~= _cmp(fp)
+   end
+
+   method gt(fp)  # self < fp?
+      return 1 = _cmp(fp)
+   end
+
+   method lt(fp)  # self > fp?
+      return -1 = _cmp(fp)
+   end
+
+   method le(fp)  # self <= fp?
+      return not gt(fp)
+   end
+
+   method ge(fp)  # self >= fp?
+      return not lt(fp)
+   end
+
+   #<p>
+   # Internal method to convert a numeric string (integer or floating point)
+   # into a digit string plus number of decimal places
+   #</p>
+   method _cnvfpt(fstr)
+      local value, places, exponent, p1
+
+      fstr ? {
+         value := if = "-" then "-" else ""
+         value ||:= tab(many(&digits)) | ""
+         if = "." then { # After decimal point
+            p1 := &pos
+            value ||:= tab(many(&digits))
+         }
+
+         if tab(any('eE')) then { # After Exponent
+            places := (&pos - \p1 - 1) | 0
+            exponent := integer(tab(0))
+         } else { # No exponent
+            places := (&pos - \p1) | 0
+            exponent := 0
+         }
+         if /exponent | not pos(0) then fail
+      }
+
+      if places > exponent then {
+         value ||:= repl("0",  0 < (exponent - places)) | ""
+      }
+      places -:= exponent
+      return VP(value,places)
+   end
+
+   # Debugging: use with {}  e.g. dbgPV{ "The values are", v1, "and", v2 }
+   method dbgPV(args)    # Remove this when the time is ripe
```

Comment:
remove?

Reply:
The dbgPV stuff is now inside $ifdef FXPT_DEBUG